File: tools/segment.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import lief


def align(value: int, page_size: int = 0x1000) -> int:
    return (value + page_size - 1) & ~(page_size - 1)


def seg_name(binary, name: str) -> str:
    if isinstance(binary, lief.MachO.Binary):
        return name if name.startswith("__") else f"__{name[:14]}"
    return name if name.startswith(".") else f".{name}"


def seg_va(binary, name: str, size: int) -> int:
    name = seg_name(binary, name)
    if isinstance(binary, lief.MachO.Binary):
        found = next((s for s in binary.segments if s.name == name), None)
        if found:
            return found.virtual_address
        last = binary.segments[-1]
        return align(last.virtual_address + last.virtual_size)
    found = binary.get_section(name)
    if found:
        return found.virtual_address
    last = binary.segments[-1]
    return align(last.virtual_address + last.virtual_size, getattr(binary, "page_size", 0x1000) or 0x1000)


@dataclass
class SegmentPlan:
    name: str
    size: int
    content: bytes = b""


def add_segment(binary_path: Path, plan: SegmentPlan, output_path: Path) -> None:
    binary = lief.parse(str(binary_path))
    if binary is None:
        raise RuntimeError(f"failed to parse {binary_path}")
    if isinstance(binary, lief.MachO.Binary):
        _add_macho(binary, plan)
    else:
        _add_elf(binary, plan)
    binary.write(str(output_path))


def write_at_offset(path: Path, offset: int, payload: bytes, size: int | None = None) -> None:
    data = bytearray(path.read_bytes())
    limit = len(payload) if size is None else size
    if offset + limit > len(data):
        raise ValueError(f"write out of range: 0x{offset:x}")
    data[offset:offset + len(payload)] = payload
    if size is not None and len(payload) < size:
        data[offset + len(payload):offset + size] = b"\x00" * (size - len(payload))
    path.write_bytes(data)


def segment_file_offset(binary, name: str) -> int:
    name = seg_name(binary, name)
    if isinstance(binary, lief.MachO.Binary):
        segment = next((s for s in binary.segments if s.name == name), None)
        if segment is None:
            raise ValueError(f"segment not found: {name}")
        section = next((s for s in binary.sections if s.segment_name == name), None)
        return section.offset if section is not None else segment.file_offset
    section = binary.get_section(name)
    if section is None:
        raise ValueError(f"section not found: {name}")
    return section.offset


def remap_macho_offset_va(before, after, file_offset: int) -> int:
    if not isinstance(before, lief.MachO.Binary) or not isinstance(after, lief.MachO.Binary):
        raise RuntimeError("failed to remap Mach-O address")
    for section in before.sections:
        start = section.offset
        end = section.offset + section.size
        if start <= file_offset < end:
            rel = file_offset - start
            found = next((s for s in after.sections if s.name == section.name and s.segment_name == section.segment_name), None)
            if found is None:
                break
            return found.virtual_address + rel
    va = before.offset_to_virtual_address(file_offset)
    if not isinstance(va, int):
        raise ValueError(f"cannot map original file offset 0x{file_offset:x}")
    return va


def _add_macho(binary, plan: SegmentPlan) -> None:
    name = seg_name(binary, plan.name)
    existing = next((s for s in binary.segments if s.name == name), None)
    content = plan.content[:plan.size].ljust(plan.size, b"\x00")
    if existing:
        existing.max_protection = 7
        existing.init_protection = 5
        existing.virtual_size = max(existing.virtual_size, align(plan.size))
        if existing.sections:
            existing.sections[0].content = memoryview(bytearray(content))
        else:
            existing.content = memoryview(bytearray(content))
        return
    last = binary.segments[-1]
    segment = lief.MachO.SegmentCommand()
    segment.name = name
    segment.virtual_address = align(last.virtual_address + last.virtual_size)
    segment.virtual_size = align(plan.size)
    segment.file_offset = align(last.file_offset + last.file_size)
    segment.max_protection = 7
    segment.init_protection = 5
    section = lief.MachO.Section()
    section.name = name
    section.segment_name = name
    section.alignment = 0
    section.content = memoryview(bytearray(content))
    segment.add_section(section)
    binary.add(segment)


def _add_elf(binary, plan: SegmentPlan) -> None:
    page = getattr(binary, "page_size", 0x1000) or 0x1000
    size = align(plan.size, page)
    name = seg_name(binary, plan.name)
    content = plan.content[:plan.size].ljust(size, b"\x00")
    section = binary.get_section(name)
    if section:
        if section.size < size:
            binary.extend(section, size - section.size)
        section.content = memoryview(bytearray(content))
        return
    section = lief.ELF.Section(name)
    section.type = lief.ELF.Section.TYPE.PROGBITS
    section.flags = lief.ELF.Section.FLAGS.ALLOC | lief.ELF.Section.FLAGS.EXECINSTR
    section.alignment = page
    section.content = memoryview(bytearray(content))
    binary.add(section, True, lief.ELF.Binary.SEC_INSERT_POS.POST_SEGMENT)

File: src/segment.cpp
#include "segment.h"
#include <cstring>
#include <cstdio>
#include <fstream>
#include <iterator>
#include <algorithm>

std::vector<uint8_t> read_file(const std::filesystem::path &path) {
    std::ifstream f(path, std::ios::binary | std::ios::ate);
    if (!f) throw std::runtime_error("cannot read: " + path.string());
    auto size = f.tellg();
    f.seekg(0);
    std::vector<uint8_t> data(size);
    if (size > 0) f.read((char *)data.data(), size);
    return data;
}

void write_file(const std::filesystem::path &path, const std::vector<uint8_t> &data) {
    std::ofstream f(path, std::ios::binary | std::ios::trunc);
    if (!f) throw std::runtime_error("cannot write: " + path.string());
    if (!data.empty()) f.write((const char *)data.data(), data.size());
}

int align(int value, int page_size) {
    return (value + page_size - 1) & ~(page_size - 1);
}

std::string seg_name(const std::string &name, bool is_macho) {
    if (is_macho) {
        if (name.size() >= 2 && name[0] == '_' && name[1] == '_')
            return name;
        return "__" + name.substr(0, 14);
    }
    if (!name.empty() && name[0] == '.')
        return name;
    return "." + name;
}

std::string seg_name(BinaryImage &binary, const std::string &name) {
    return seg_name(name, binary.is_macho());
}

static uint64_t last_end_va(const std::vector<BinarySegment> &segments,
                            uint64_t default_page = 0x1000) {
    uint64_t end = 0;
    for (auto &segment : segments)
        end = std::max(end, segment.virtual_address + segment.virtual_size);
    return end > 0 ? (end + default_page - 1) & ~(default_page - 1) : 0;
}

uint64_t seg_va(BinaryImage &binary, const std::string &name, int size) {
    (void)size;
    if (binary.is_macho()) {
        auto sname = seg_name(name, true);
        for (auto &segment : binary.segments())
            if (segment.name == sname) return segment.virtual_address;
        return last_end_va(binary.segments());
    }
    if (binary.is_elf()) {
        auto sname = seg_name(name, false);
        auto *section = binary.section(sname);
        if (section) return section->virtual_address;
        return last_end_va(binary.segments());
    }
    throw std::runtime_error("unsupported binary type");
}

void add_segment(const std::filesystem::path &binary_path,
                 const SegmentPlan &plan,
                 const std::filesystem::path &output_path) {
    auto binary = BinaryImage::parse(binary_path);
    if (!binary)
        throw std::runtime_error("failed to parse " + binary_path.string());
    int size = plan.size;
    auto content = plan.content;
    content.resize(size, 0);
    binary->add_executable_section(seg_name(*binary, plan.name), size, content);
    binary->write(output_path);
}

void write_at_offset(const std::filesystem::path &path, int offset,
                     const std::vector<uint8_t> &payload, int size) {
    auto data = read_file(path);
    int limit = size >= 0 ? size : (int)payload.size();
    if (offset + limit > (int)data.size())
        throw std::runtime_error("write out of range");
    memcpy(&data[offset], payload.data(), payload.size());
    if (size > (int)payload.size())
        memset(&data[offset + payload.size()], 0, size - payload.size());
    write_file(path, data);
}

int segment_file_offset(BinaryImage &binary, const std::string &name) {
    if (binary.is_macho()) {
        auto sname = seg_name(name, true);
        for (auto &section : binary.sections())
            if (section.segment_name == sname) return (int)section.offset;
        for (auto &segment : binary.segments())
            if (segment.name == sname) return (int)segment.file_offset;
        throw std::runtime_error("segment not found: " + name);
    }
    if (binary.is_elf()) {
        auto sname = seg_name(name, false);
        auto *section = binary.section(sname);
        if (!section) throw std::runtime_error("section not found: " + name);
        return (int)section->offset;
    }
    throw std::runtime_error("unsupported binary type");
}

uint64_t remap_macho_offset_va(BinaryImage &before,
                                BinaryImage &after,
                                int file_offset) {
    for (auto &sec : before.sections()) {
        int start = (int)sec.offset;
        int end = start + (int)sec.size;
        if (start <= file_offset && file_offset < end) {
            int rel = file_offset - start;
            for (auto &s : after.sections()) {
                if (s.name == sec.name && s.segment_name == sec.segment_name)
                    return s.virtual_address + rel;
            }
            break;
        }
    }
    auto va = before.offset_to_virtual_address(file_offset);
    if (!va)
        throw std::runtime_error("cannot map original file offset");
    return *va;
}
